### PR TITLE
Refactor string interpolators and error message handling

### PR DESCRIPTION
**Pull request by odersky**
The idea is that we want to concentrate error handling logic in the `Message` type instead of dispersing it over various
string interpolators.

 - [x] Pass more messages instead of strings to `report` methods.
 - [x] Make this easier by having string interpolators that produce messages
 - [x] Use messages instead of strings elsewhere
 - [x] Move logic for limiting size of messages and filtering out nonsensical errors into `Message`.

From the doc comment for `Message`:

## Tips for error message generation

 - You can use the `em` interpolator for error messages. It's defined in core.Decorators.
 - You can also use a simple string argument for `error` or `warning` (not for the other variants),
   but the string should not be interpolated or composed of objects that require a
   Context for evaluation.
 - When embedding interpolated substrings defined elsewhere in error messages,
   use `i` and make sure they are defined as def's instead of vals. That way, the
   possibly expensive interpolation will performed only in the case where the message
   is eventually printed. Note: At least during typer, it's common for messages
   to be discarded without being printed. Also, by making them defs, you ensure that
   they will be evaluated in the Message context, which makes formatting safer
   and more robust.
 - For common messages, or messages that might require explanation, prefer defining
   a new `Message` in file `messages.scala` and use that instead. The advantage is that these
   messages have unique IDs that can be referenced elsewhere.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Decorators.scala [hunk 1]
```diff
@@ -291,17 +291,41 @@ object Decorators {
     def derivedCons(x1: T, xs1: List[T]) =
       if (xs.head eq x1) && (xs.tail eq xs1) then xs else x1 :: xs1
 
+  /* This extension defines the following string formatters:
+   *   i, e, ex, em, exm
+   * String formatters ending in `m` produce a message, the others produce a string.
+   * Details are as follows:
+   *
+   *   i  : Gneral purpose string interpolator that uses Formtting.show to print compiler data
```
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/core/Decorators.scala:299]:
```suggestion
   *   i  : General purpose string interpolator that uses Formatting.show to print compiler data
```

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Decorators.scala:299]: There will be more to come. Hopefully, we can eliminate the zoo of string interpolators and push much more functionality into `Message`.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/Decorators.scala:301]: Sounds good to me!

**Comment by odersky** [comment]
Who would like to give this a review? To be sure, given the change set a lof of it will have to be cursory.

**Review by dwijnand — APPROVED** [review]
I've been reading the diffs as you've been pushing them.  The only thing I didn't reread when code moved files, so I missed any changes in there (if any).  But the rest LGTM.

**Comment by odersky** [comment]
OK, great! The code that moved from Formatting to Message was unchanged except for straightforward refactorings.

**Pull request merged**